### PR TITLE
Sign In Feature

### PR DESCRIPTION
**Pull request by markquark125**
Check README in -the account- folder. Also, because I use a Windows environment, I had to change a script in package.json

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,14 +9,16 @@
...
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
+    "react-native": "^0.71.8",
     "react-router-dom": "^6.11.1",
     "react-scripts": "5.0.1",
     "web-vitals": "^2.1.4"
   },
   "scripts": {
-    "start": "HTTPS=true SSL_CRT_FILE=./ssl-certificates/localhost+2.pem SSL_KEY_FILE=./ssl-certificates/localhost+2-key.pem react-scripts start",
+    "start": "set HTTPS=true && set SSL_CRT_FILE=./ssl-certificates/localhost.pem && set SSL_KEY_FILE=./ssl-certificates/localhost-key.pem && react-scripts start",
```
Comment (qingquan-li) [inline package.json]: Did you change this for your Windows environment? You can change it locally; however, you should change it back before using git to commit it.

File: src/App.js [hunk 2]
```diff
@@ -8,16 +8,19 @@ import BottomNavBar from "./common/BottomNavBar";
 import Home from "./home";
 import Navigation from './navigation';
 import Account from "./account";
+import { useAuthentication } from "./common/useAuthentication";
 
 function App() {
+  const {user} = useAuthentication();
```
Comment (qingquan-li) [inline src/App.js]: I did not understand this line (please add the necessary comments). You did not use the user object here.

File: src/App.js [hunk 3]
```diff
@@ -8,16 +8,19 @@ import BottomNavBar from "./common/BottomNavBar";
...
+
   return (
     <Router>
       <div className="flex flex-col min-h-screen">
         <div className="flex-grow">
           <Routes>
             <Route path="/" element={<Home />} />
             <Route path="/navigation" element={<Navigation />} />
-            <Route path="/account" element={<Account />} />
+            <Route path="/account" element={<Account/>}/>
```
Comment (qingquan-li) [inline src/App.js]: Please pay attention to the code specification. Why did you delete the spaces?

File: src/account/README.MD [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+- Tried conditional rendering to render either sign-in or sign-up form. Uses `handleLoginClick` for conditional rendering.
+- call `useAuthentication()` in `App.js` to initialize, get Auth instance, and set an observer with Firebase for user sign-in state.
+
+---
+
+# Dev Issues
+
+## Firebase options
+
+- I kept getting an error like `Firebase: Need to provide options, when not being deployed to hosting via source.` Not sure why this happens. I already called `useAuthentication` in `App.js` to initialized for Firebase.
```
Comment (qingquan-li) [inline src/account/README.MD]: This error suggests that your Firebase application needs some configuration options that are not being provided. You did not export the configuration in `useAuthentication.js` correctly.

File: src/account/SignIn.jsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
+
+import React from 'react';
+import { getAuth, signInWithEmailAndPassword } from 'firebase/auth';
+
+
+const auth = getAuth();
```
Comment (qingquan-li) [inline src/account/SignIn.jsx:5]: You may need to do something like this here:

```
import { initializeApp } from "firebase/app";
const firebaseConfig = {
  apiKey: "your-api-key",
  authDomain: "your-auth-domain",
  projectId: "your-project-id",
  storageBucket: "your-storage-bucket",
  messagingSenderId: "your-messaging-sender-id",
  appId: "your-app-id"
};
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
```

File: src/common/useAuthentication.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// Left as TypeScript. Can be refactored into JavaScript but this is fine for demo purposes.
+// Added for Firebase Authorization. Decided to leave config in for refactoring(Mark)
+
+import React from 'react';
+import { initializeApp } from "firebase/app";
+import { getAuth, onAuthStateChanged, User} from "firebase/auth";
+
+// import Constants from 'expo-constants';
```
Comment (qingquan-li) [inline src/common/useAuthentication.js]: Please delete useless comments. They may confuse other collaborators.

File: src/common/useAuthentication.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
+// Left as TypeScript. Can be refactored into JavaScript but this is fine for demo purposes.
```
Comment (qingquan-li) [inline src/common/useAuthentication.js]: This file should be placed in the "account" folder.

**Review by qingquan-li — CHANGES_REQUESTED** [review]

**Comment by qingquan-li** [comment]
@markquark125 Please check my code review comments and make some changes.
You can try to follow tutorials or youtube videos to develop authentication functionality.
BTW, never create a pull request to merge your branch until it works. Please check the `CONTRIBUTING.md` file.

File: package.json (hunk 1 above)
Comment (markquark125) [reply package.json]: Yes, I had to change for Windows environment. I'll change it before pushing again

File: src/common/useAuthentication.js (hunk 7 above)
Comment (markquark125) [reply src/common/useAuthentication.js]: Will do

File: src/common/useAuthentication.js (hunk 6 above)
Comment (markquark125) [reply src/common/useAuthentication.js]: Noted

File: src/App.js (hunk 2 above)
Comment (markquark125) [reply src/App.js]: Thank you. I think I see the issue with the Firebase authentication.

File: src/App.js (hunk 3 above)
Comment (markquark125) [reply src/App.js]: Whoops. Will revert to the original.

File: src/account/README.MD (hunk 4 above)
Comment (markquark125) [reply src/account/README.MD]: Noted.

File: src/account/SignIn.jsx (hunk 5 above)
Comment (markquark125) [reply src/account/SignIn.jsx:5]: Figure out the issue. It was an issue with importing and moving certain files

File: .gitignore [hunk 8]
```diff
@@ -17,6 +17,7 @@
 .env.development.local
 .env.test.local
 .env.production.local
+.env
```
Comment (qingquan-li) [inline .gitignore:20]: Where do you use it? If you use the `dotenv` library, please explain how to handle the variables in the README.

File: src/account/README.MD [hunk 9]
```diff
@@ -1,26 +1,62 @@
...
+## ERROR: Dev Issues
+
+### Firebase options
+
+- I kept getting an error like `Firebase: Need to provide options, when not being deployed to hosting via source.` Not sure why this happens. I already called `useAuthentication` in `App.js` to initialized for Firebase.
+
 ---
 
-# Dev Issues
+# Sign-in feature is functional (notes, 5/12/2023)
```
Comment (qingquan-li) [inline src/account/README.MD]: Nice summary below. But you should put this changelog on the git commit.

File: package.json [hunk 10]
```diff
@@ -9,14 +9,16 @@
     "@testing-library/jest-dom": "^5.16.5",
     "@testing-library/react": "^13.4.0",
     "@testing-library/user-event": "^13.5.0",
+    "firebase": "^9.21.0",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
+    "react-native": "^0.71.8",
```
Comment (qingquan-li) [inline package.json]: We never use `react-native` in this app!! Please do a code review by yourself before pushing the code.

File: src/account/SignIn.jsx [hunk 11]
```diff
@@ -1,20 +1,16 @@
-
 import React from 'react';
-import { getAuth, signInWithEmailAndPassword } from 'firebase/auth';
-
+import { getAuth, signInWithEmailAndPassword} from 'firebase/auth';
 
 const auth = getAuth();
 
 export default function SignInScreen(){
```
Comment (qingquan-li) [inline src/account/SignIn.jsx]: If it is a component, you can name it `SignIn` instead of `SignInScreen`.

File: src/account/SignUp.jsx [hunk 12]
```diff
@@ -29,7 +28,7 @@ export default function SignUpScreen(props) {
     }) */
     try {
       await createUserWithEmailAndPassword(auth, value.email, value.password);
-      props.SetSignUp(true);
+      alert("Registered!");
```
Comment (qingquan-li) [inline src/account/SignUp.jsx]: Do you mean "Sign Up Successfully"? Please keep the wording consistent.

File: src/account/SignUp.jsx [hunk 13]
```diff
@@ -44,30 +43,34 @@ export default function SignUpScreen(props) {
 
     <div>
       <h1>Sign Up Screen!</h1>
```
Comment (qingquan-li) [inline src/account/SignUp.jsx:43]: You did not navigate the user to the (new) Sign Up page. The user is still on the account page. When a user clicks the Sign Up button, you should navigate the user to the Sign Up page.

**Review by qingquan-li — CHANGES_REQUESTED** [review]
I can not sign up successfully. The Cloudflare Pages build failed, either. Please fix it.
Please Todo:
- Please add more comments for your code; for example, add necessary comments for a function to explain how it works.
- Build user-friendly UI. For example, when a user sign up fail, tell the user why and how to fix it. Please keep the wording consistent: don't use `sign up` and `register` together, and don't `sign in` and `log in` together. Never confuse users.
- Form validation. For example, if the user enters "username@gmail" in the email input box, the user should be reminded that the email format is wrong. If you have time, you need to verify that the user's email address is real and valid in the form of a verification code.
- When a user wants to submit accessible storefront data to our database, we must check if the user logs in. How can we check the login state of a user? Please explain it in your README file.

File: package.json (hunk 10 above)
Comment (markquark125) [reply package.json]: will delete

File: src/account/SignIn.jsx (hunk 11 above)
Comment (markquark125) [reply src/account/SignIn.jsx]: changed it

File: src/account/SignUp.jsx (hunk 12 above)
Comment (markquark125) [reply src/account/SignUp.jsx]: changed it

File: src/account/SignUp.jsx (hunk 13 above)
Comment (markquark125) [reply src/account/SignUp.jsx:43]: I'm not sure what you mean. When you click on the button, it renders the Sign Up page. Similar logic as the Sign In page.

Comment (qingquan-li) [reply src/account/SignUp.jsx:43]: For example, a user clicks the Sign Up button and should be navigated to a new page (Sign Up page). The user can enter his/her email and password to sign up on this new page.
It will be better to create a new page to handle user sign up / sign in. Because in the future, we may add "reset password," "forgot password," "enter verification code," ... features/components on the Sign Up/Sign In page.
Also, we will add more features / components on the Account page. So, please do not render the sign up / sign in form on the Account page. Create new pages instead.

File: package.json (hunk 1 above)
Comment (qingquan-li) [reply package.json]: You changed this configuration again. Please do some review before pushing code.

**Review by qingquan-li — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (markquark125) [reply package.json]: Oh, whoops. I kept having to change it for me to test. Will fix and push again.

File: src/account/useAuthentication.js [hunk 14]
```diff
@@ -5,31 +5,32 @@ import { getAuth, onAuthStateChanged} from "firebase/auth";
...
-    apiKey: "PLACEHOLDER",
-    authDomain: "PLACEHOLDER",
-    projectId: "PLACEHOLDER",
-    storageBucket: "PLACEHOLDER",
-    messagingSenderId: "PLACEHOLDER",
-    appId: "PLACEHOLDER"
+// Your app's Firebase configuration (HARDCODED. NEEDS REVISION)
+// Left below as a placeholder.
+  const firebaseConfig = {
+    apiKey: process.env.FIREBASE_API_KEY,
```
Comment (qingquan-li) [inline src/account/useAuthentication.js]: Awesome to use `process.env`.
Note: In a Create React App project, environment variables need to start with `REACT_APP_` to be embedded into the (Cloudflare Pages) build.
Please change to:
```
const firebaseConfig = {
  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.REACT_APP_FIREBASE_APP_ID
};
```
Please delete `databaseURL` because we use MongoDB instead of the firebase database.

**Review by qingquan-li — CHANGES_REQUESTED** [review]
I can run your code locally now. Well done. 👍
But your code cannot build and run successfully on Cloudflare Pages. Please fix it following my code review comments.

File: src/account/useAuthentication.js (hunk 14 above)
Comment (markquark125) [reply src/account/useAuthentication.js]: Will change and push. thanks!

Comment (qingquan-li) [reply src/account/useAuthentication.js]: Also, you should request to merge your branch into `dev` instead of `main` branch. Please check the `CONTRIBUTING.md` file.

Comment (qingquan-li) [reply src/account/useAuthentication.js]: > Also, you should request to merge your branch into `dev` instead of `main` branch. Please check the `CONTRIBUTING.md` file.

@markquark125 You don't need to create a new pull request to change to merge your `#account` branch into the `dev` branch. Please click the `Edit` button on the top-right to change the base branch from `main` to `dev`.

Comment (markquark125) [reply src/account/useAuthentication.js]: Edited the pull request so #account -> dev

**Review by qingquan-li — APPROVED** [review]
